File: matlab/Lyapunov_f.m
function g=Lyapunov_f(y,m,D);   %Оценивание показателя Ляпунова
%по отрезку ряда y для D контрольных шагов с m зонами разбиения
%clear; clc;                  %показатель хаотичности по Ляпунову           
%load('Data','YY');  [N,m]=size(YY);   % <288000x6>     6 каналов
%L=200;  r=5;  y=YY(1:L,r);      %берём канал номер r
%m=30;   D=4;
    L=length(y);
    eps=(max(y)-min(y))/m;          % высота ячейки разбиения

    nn=1:3:round(L/2);
for n=1:3:round(L/2);
    y0=y(n); I=find(abs(y-y0)<eps);  P=length(I);
    v=ones(1,length(nn)); T=ones(1,length(nn));
    for k=1:P-1;
       z1=y(I(k):L-(I(k+1)-I(k)));  z2=y(I(k+1):L);   z=abs(z1-z2);
       JJ=find(z>D*eps);       J=min(min(JJ));
       u1=abs(z1(1)-z2(1)+1);  u2=abs(z1(J)-z2(J)+1);
          if (u1~=0)&(u2~=0);  
             v(k)=log(u2+10)-log(u1+10);    T(k)=J;
          else v(k)=0;  T(k)=0;
          end;        
   end;
   V(n)=sum(v);   TT(n)=sum(T);
end;
g=sum(V)/sum(TT);
    

File: functions/Lyapunov_f.py
"""
Оценивание показателя Ляпунова
"""
import numpy as np


def Lyapunov_f(y, m, D):
    """
    Вычисление показателя Ляпунова

    Parameters:
    -----------
    y : numpy.ndarray
        Входной ряд данных
    m : int
        Число зон разбиения
    D : float
        Отклонение для контрольных шагов

    Returns:
    --------
    g : float
        Показатель Ляпунова
    """
    y = np.array(y).flatten()
    L = len(y)

    if L < 2:
        return 0.0

    eps = (np.max(y) - np.min(y)) / m

    if eps == 0:
        return 0.0

    nn = list(range(1, round(L/2) + 1, 3))
    V = np.zeros(len(nn))
    TT = np.zeros(len(nn))

    for idx, n in enumerate(nn):
        y0 = y[n-1]  # Индексация с 0 (MATLAB: y(n))
        I = np.where(np.abs(y - y0) < eps)[0]
        P = len(I)

        v = np.zeros(P - 1)
        T = np.zeros(P - 1)

        for k in range(P - 1):
            idx1 = I[k]
            idx2 = I[k + 1]
            diff = idx2 - idx1

            if idx1 + diff < L and idx2 < L:
                z1 = y[idx1:L - diff]
                z2 = y[idx2:L]
                z = np.abs(z1 - z2[:len(z1)])

                JJ = np.where(z > D * eps)[0]

                if len(JJ) > 0:
                    J = np.min(JJ)

                    if J < len(z1) and J < len(z2):
                        u1 = abs(z1[0] - z2[0] + 1)
                        u2 = abs(z1[J] - z2[J] + 1)

                        if u1 != 0 and u2 != 0:
                            v[k] = np.log(u2 + 10) - np.log(u1 + 10)
                            T[k] = J

        V[idx] = np.sum(v)
        TT[idx] = np.sum(T)

    if np.sum(TT) > 0:
        g = np.sum(V) / np.sum(TT)
    else:
        g = 0.0

    return g

